### PR TITLE
Add context menu for locator filter results

### PR DESCRIPTION
**Pull request by 3nids**
This adds the possibility to add context menu for locator results
Opening form option has been added for all layers search filter

image

### CODE REVIEW

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 1]
```diff
@@ -400,6 +401,8 @@ void QgsAllLayersFeaturesLocatorFilter::fetchResults( const QString &string, con
       result.userData = QVariantList() << f.id() << preparedLayer.layerId;
       result.icon = preparedLayer.layerIcon;
       result.score = static_cast< double >( string.length() ) / result.displayString.size();
+
+      result.contextMenuActions.insert( OpenForm, new QAction( tr( "Open form" ) ) );
```
Comment (3nids) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp]: @nyalldawson I guess there is a leak there. I am not sure how to handle this. Is it the responsibility of the model to delete the actions when it's cleared?

Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp]: "Open Form..."

File: src/core/locator/qgslocatorfilter.h [hunk 2]
```diff
@@ -87,10 +89,20 @@ class CORE_EXPORT QgsLocatorResult
...
 
+    /**
+      * Actions to be used in a context menu for the result.
+      * The key of the map is populated with IDs used to recognized
+      * entry when the result is triggered. The IDs should be 0 or greater
+      * otherwise, the result will be triggered normally.
+      * Entries in the context menu will be ordered by IDs.
+      * \since QGIS 3.6
+      */
+    QMap<int, QAction *> contextMenuActions = QMap<int, QAction *>();
```
Comment (nyalldawson) [inline src/core/locator/qgslocatorfilter.h]: Can we replace this with the more generic "actions" name? I'm thinking specifically for QField and other clients, where the results may not be shown in a context menu but in some other way.

Comment (nyalldawson) [reply src/core/locator/qgslocatorfilter.h]: (There's also no need to initialise this list here)

File: src/core/locator/qgslocatorfilter.h [hunk 3]
```diff
@@ -209,6 +221,14 @@ class CORE_EXPORT QgsLocatorFilter : public QObject
      */
     virtual void triggerResult( const QgsLocatorResult &result ) = 0;
 
+    /**
+     * Triggers a filter \a result from this filter for an entry in the context menu.
+     * The entry is identified by its \id as specified in the result of this filter.
+     * \see triggerResult()
+     * \since QGIS 3.6
+     */
+    virtual void triggerResultFromContextMenu( const QgsLocatorResult &result, const int id );
```
Comment (nyalldawson) [inline src/core/locator/qgslocatorfilter.h]: Same here -- rename to "triggerResultFromAction"?

**Review by nyalldawson — COMMENTED** [review]
I like the direction you're taking with this, but I'm wondering if relying on QAction is a mis-step. We may be better served here with a simple little "ResultAction" struct, with members like int ID and QString text. 

This would allow us to move all the QAction code out of core, and only construct QActions when building the menu in QgsLocatorWidget::showContextMenu. A benefit here is that the parent menu is known at this stage, so the actions can be correctly parented to the menu and avoiding the leak you described.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Looks perfect!

**Pull request merged**